File: base/data/intcount.cpp
/********************************************************************
 *
 *  intcount.cpp
 *
 *  The C++ mex implementation of integer counting
 *
 *  Created by Dahua Lin, on May 26, 2010
 *
 ********************************************************************/

#include <bcslib/matlab/bcs_mex.h>

using namespace bcs;
using namespace bcs::matlab;

template<typename T>
inline void get_vs(const_marray mRgn, int& v0, int& v1)
{
    size_t ne = mRgn.nelems();
    const T *v = mRgn.data<T>();
        
    if (ne == 1)
    {
        v0 = 1;
        v1 = (int)(*v);
    }
    else
    {
        v0 = (int)(v[0]);
        v1 = (int)(v[1]);
    }
}



inline void get_range(const_marray mRgn, int& v0, int& v1)
{
    size_t ne = mRgn.nelems();
    
    if (!( (ne == 1 || ne == 2) && !mRgn.is_sparse() ) )
    {
        throw mexception("intcount:invalidarg", 
            "The range [v0 v1] should be a (non-sparse) scalar or pair.");
    }
    
    if (mRgn.is_double())
    {
        get_vs<double>(mRgn, v0, v1);
    }
    else if (mRgn.is_single())
    {
        get_vs<float>(mRgn, v0, v1);
    }
    else if (mRgn.is_int32())
    {
        get_vs<int32_t>(mRgn, v0, v1);
    }
    else
    {
        throw mexception("intcount:invalidarg", 
            "The range [v0 v1] should be of class double, single, or int32");
    }
}


template<typename T>
void count(int v0, int v1, const T *v, size_t n, double *c)
{
    for (size_t i = 0; i < n; ++i)
    {
        int cv = (int)(v[i]);
        if (cv >= v0 && cv <= v1)
        {
            ++ c[cv - v0];
        }
    }
}


inline marray do_count(int v0, int v1, const_marray mVals)
{
    int m = v1 - v0 + 1;
    marray mCount = create_marray<double>(1, m);
    double *c = mCount.data<double>();
    
    size_t n = mVals.nelems();
    
    if (mVals.is_double())
    {
        count(v0, v1, mVals.data<double>(), n, c);
    }
    else if (mVals.is_single())
    {
        count(v0, v1, mVals.data<float>(), n, c);
    }
    else if (mVals.is_int32())
    {
        count(v0, v1, mVals.data<int32_t>(), n, c);
    }
    else
    {
        throw mexception("intcount:invalidarg", 
            "The class of values should be either double, single, int32");
    }
    
    return mCount;
    
}


/***********
 *
 *  Main entry
 *   
 *  Input:
 *    [0]:  the value range in form of [v0, v1] [pair (double|single|int32)]
 *    [1]:  the array of values to count [array (double|single|int32)]
 *
 *  Output:
 *    [0]:  the vector of counts (of size 1 x (v1 - v0 + 1)) [double array]
 *
 *  No 
 */
void bcsmex_main(int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[])
{
    if (nrhs != 2)
    {
        throw mexception("intcount:invalidarg", 
            "The number of inputs to intcount should be 2.");
    }

    const_marray mRgn(prhs[0]);
    const_marray mVals(prhs[1]);

    int v0, v1;
    get_range(mRgn, v0, v1);

    plhs[0] = do_count(v0, v1, mVals).mx_ptr();

}


BCSMEX_MAINDEF

File: base/data/intcount.m
function c = intcount(rgn, V, dim)
%INTCOUNT Count the number of occurrence of integers
%
%   c = INTCOUNT([v0, v1], V);
%       counts the number of integers in [v0, v1] that occurs in V.
%       
%       In the input, V is an array containing integer values (whose
%       type can be double or single nonetheless). 
%       In the output, c is a row vector of length v1 - v0 + 1.
%       In particular, c(i) counts the number of times that the value
%       v0 + (i-1) occurs in V.
%
%   c = INTCOUNT(K, V);
%       This is equivalent to intcount([1, K], V);
%
%   c = INTCOUNT([v0, v1], V, dim);
%   c = INTCOUNT(K, V, dim);
%
%       performs the counting along specified direction. Here, dim
%       can be either 1 or 2. 
%
%       If dim == 1, it counts for each column, resulting a vector
%       of counts, whose size is 1 x n.
%
%       If dim == 2, it counts for each row, resulting a vector of 
%       counts, whose size is m x 1.
%

%   History
%   -------
%       - Created by Dahua Lin, on May 26, 2010
%       - Modified by Dahua Lin, on Jun 6, 2010
%           - Use pure C++ mex implementation
%       - Modified by Dahua Lin, on Feb 26, 2012
%           - Support per column/row counting
%

%% verify input arguments

if isnumeric(rgn) && isvector(rgn) 
    if isscalar(rgn)
        v0 = 1;
        v1 = rgn;
        if v1 < 1
            error('intcount:invalidarg', 'K should be a positive number.');
        end        
    elseif numel(rgn) == 2
        v0 = rgn(1);
        v1 = rgn(2);
        if v0 > v1
            error('intcount:invalidarg', 'The condition v0 <= v1 is not met.');
        end        
    else
        error('intcount:invalidarg', 'The first argument is invalid.');
    end
else
    error('intcount:invalidarg', 'The first argument is invalid.');
end
v0 = int32(v0);
v1 = int32(v1);

if ~(isnumeric(V) && isreal(V) && ~issparse(V))
    error('intcount:invalidarg', ...
        'V should be a non-sparse numeric array of integers.');
end

if nargin < 3
    dim = 0;
else
    if ndims(V) > 2
        error('intcount:invalidarg', ...
            'V need to be a matrix when dim is explicitly specified.');
    end    
    if isequal(dim, 1)
        dim = 1;
    elseif isequal(dim, 2)
        dim = 2;
        V = V.';
    end
end

%% main

c = intcount_cimp(v0, v1, V, dim > 0);

if dim == 2
    c = c.';
end

    